talk/expert/modules.tex

```diff
@@ -0,0 +1,510 @@
+      \begin{itemize}
+        \item E.g. \mintinline{cpp}{memcpy(a, b, n);}, \mintinline{cpp}{sqrt(val);}, \mintinline{cpp}{uint32_t i;} will not compile. You need to prefix \mintinline{cpp}{std::}.
+      \end{itemize}
+      \item \mintinline{cpp}|import std.compat;| for the entire C and C++ standard library.
+      \item No macros are exported. For these, you still need to include the corresponding headers.
+      \begin{cppcode*}{gobble=4}
+        import std;        // everything C++
+        #include <cassert> // for assert()
+        #include <climits> // for e.g. CHAR_BIT
+      \end{cppcode*}
+      \item Further modularization of the standard library is in progress.
+    \end{itemize}
+  \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Build system}
+    \begin{block}{Challenges}
+      \begin{itemize}
+        \item Resolving a module name to the file name building it
+        \item Translation units no longer independent
+        \begin{itemize}
+          \item Extra tool needs to infer dependency graph before building
+          \item This graph needs to be communicated to the build system
+          \item We likely need a \href{https://wg21.link/P1689}{standard dependency file format}
+          \item Parallel and distributed builds need synchronization
+        \end{itemize}
+        \item Compilation of module translation units produces binary module interface (BMI) and object file
+        \begin{itemize}
+          \item BMIs need to be maintained/shared between multiple compiler instances
+        \end{itemize}
+        \item Tools beside the compiler need to build/read binary module interface (static analysis, auto completion, etc.)
+        \item The C++ standard specifies very little on how this works
+        \begin{itemize}
+          \item We may experience large implementation divergence
+        \end{itemize}
+      \end{itemize}
+    \end{block}
+\end{frame}
+
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Build system}
+    \begin{block}{Status}
+      \begin{itemize}
+        \item build2 supports everything g++ does
+        \item MSBuild figures out dependencies and build order automatically (Windows/Visual Studio only)
+        \item \mintinline{cpp}{g++ a.cpp b.cpp ...} "just works", must be one line
+        \item GCC uses a \href{https://wg21.link/P1184}{module mapper} to resolve imports, which specifies a protocol and uses a central server for module maintenance
+        \item \href{https://wg21.link/p1602}{Experimental extensions to GNU make} to grow dependency graph on the fly while modules are discovered
+        \item No support in cmake (3.23) yet
+      \end{itemize}
+    \end{block}
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Building modules (g++)}
+    \begin{block}{Case study: g++}
+      \begin{itemize}
+        \item By default, g++ caches compiled module interfaces (CMI, i.e.\ BMI) in subdirectory \texttt{./gcm.cache}
```

Comment:
Bravo!

Reply:
Because I put the backslash in `i.e.\` myself? :D